### PR TITLE
fix(terrain): consistent terrain status across legend + settings

### PR DESCRIPTION
**Pull request by ishiland**
Two related fixes making the terrain truth consistent across the map, the legend, and the settings panel. Both stem from the renderer resolving terrain **mode-agnostically** (a hillshade-mode DEM drives the 3D mesh and paints its 2D relief) while other surfaces hadn't caught up.

## 1. Legend: dedup the synthetic "3D terrain" entry
On the swissALTI3D Matterhorn map the on-map legend showed a "3D terrain" row with no counterpart in the layer list that **duplicated** the visible "swissALTI3D relief" hillshade — same dataset (`terrain_config.source_dataset_id` == the visible DEM's `dataset_id`). `deriveTerrainLegendEntry` fires render-mode-agnostically, so when the source DEM also paints a visible hillshade the dataset appears twice.

Fix: suppress the synthetic entry when the terrain source dataset is already a **visible** per-layer entry, via a shared `terrainSourceIsShownAsLayer` gated in both consumers (builder `LegendPlugin` against `legendLayers`, viewer `LayerLegend` against currently-visible `sorted` entries — the viewer keeps toggled-off rows, so it must dedup against visible ones only, else toggling the hillshade off would hide both it and the synthetic row while terrain is still active). Kept for the pure-terrain / hidden-source case where the synthetic row is the only indicator.

## 2. Settings: "No terrain layer is active" shown while terrain renders
The map renderer (`BuilderMap` FIX-3-RESOLVER D-06) resolves the terrain DEM by `source_dataset_id` + `isTerrainCapableDemLayer`, **not** `render_mode === 'terrain'`. But `MapBuilderPage.isTerrainActive` (via `boundTerrainLayer`) still required terrain mode — so the Matterhorn map (terrain enabled, DEM in hillshade mode) reported **"No terrain layer is active"** while the 3D mesh was rendering from that DEM.

Fix: extract one shared `resolveTerrainSourceLayer` (map-stack) and consume it in **both** the renderer and the status computation so they can't drift again (the drift is what caused this bug). `isTerrainActive` now also honors the DEM's visibility, matching the renderer's `effectiveTerrainEnabled` (terrain clears when the bound DEM is hidden). Stale `SettingsEditorScene` JSDoc updated.

## Tests
- `terrain-legend.test.ts` — `terrainSourceIsShownAsLayer` unit tests
- `LegendPlugin.terrain.test.tsx` + `LayerLegend.test.tsx` — dedup guards incl. the toggled-off-source case (Codex P2 regression)
- `map-stack.test.ts` — `resolveTerrainSourceLayer` unit tests (hillshade + terrain modes resolve; non-DEM / wrong dataset / no source → undefined)
- Existing D-01/D-02/D-03, position, dangling, and BuilderMap terrain-visibility cases unchanged. Typecheck + lint clean.

## Out of scope (noted)
The builder *layer list* (`UnifiedStackPanel`) renders raw layers and has no terrain row — after fix 1 the legend matches it. The stack's `makeTerrainReliefEntry` still computes terrain "active" by terrain-mode only, but that row isn't rendered anywhere user-facing.

### CODE REVIEW

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/viewer/LayerLegend.tsx [hunk 1]
```diff
@@ -211,10 +212,15 @@ export function LayerLegend({
...
-  const terrainEntry = useMemo(
-    () => deriveTerrainLegendEntry(terrainConfig, layers, { labelKey: 'viewer.legend.terrain3d' }),
-    [terrainConfig, layers],
-  );
+  const terrainEntry = useMemo(() => {
+    const entry = deriveTerrainLegendEntry(terrainConfig, layers, { labelKey: 'viewer.legend.terrain3d' });
+    // Dedup: drop the synthetic entry when the terrain source DEM is also shown
+    // as a per-layer entry (e.g. a visible hillshade of the same dataset), so
+    // the legend doesn't list one DEM twice. Kept for the pure-terrain case.
+    return entry && !terrainSourceIsShownAsLayer(terrainConfig, sorted.map((s) => s.layer))
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/viewer/LayerLegend.tsx]: Good catch — fixed in 40beb297. The viewer now dedups against currently-visible entries only (`sorted.filter(s => visibleLayers.has(s.key))`) with `visibleLayers` added to the memo deps, so toggling the source hillshade off keeps the synthetic 3D-terrain row. Added a regression test for that case. The builder LegendPlugin was unaffected (its `legendLayers` already filters on `l.visible`).

**Comment by ishiland** [comment]
@codex review

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/viewer/__tests__/LayerLegend.test.tsx [hunk 2]
```diff
@@ -344,4 +344,57 @@ describe('LayerLegend terrain consistency (Fix 1)', () => {
...
+        onToggle={vi.fn()}
+      />,
+    );
+
+    expect(screen.getByText('swissALTI3D relief')).toBeInTheDocument();
+    expect(screen.queryByTestId('legend-terrain-synthetic')).not.toBeInTheDocument();
+    expect(screen.queryByText('3D terrain')).not.toBeInTheDocument();
+  });
+
+  // Codex P2: the viewer keeps toggled-off layers in the list (visibility lives
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/viewer/__tests__/LayerLegend.test.tsx]: Done in c3e4880d — replaced the `// Codex P2:` tag with the codebase's standard `Regression:` idiom (already used elsewhere in this file). Rationale is stated inline, no un-lookup-able finding id.

**Comment by ishiland** [comment]
Added a second related fix (979e7a81): the settings panel reported "No terrain layer is active" while the map rendered terrain from a hillshade-mode DEM. Extracted a shared `resolveTerrainSourceLayer` so the renderer and the settings status use one resolver. @codex review

File: frontend/src/components/builder/BuilderMap.tsx [hunk 3]
```diff
@@ -40,11 +40,11 @@ import {
   getLayerId,
   ensureRasterDemTerrainSource,
   clearTerrainForStyleSwap,
-  isTerrainCapableDemLayer,
   normalizeTerrainExaggeration,
   refreshVectorSourceTiles,
   TERRAIN_SOURCE_ID,
 } from './map-sync';
+import { resolveTerrainSourceLayer } from './map-stack';
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/builder/BuilderMap.tsx:47]: Correct, and a serious catch — thank you. A botched `git add` (wrong test path aborted the whole command under `2>/dev/null`) committed the imports + unit test but dropped `map-stack.ts` itself, so the export was missing and the branch failed TS2305. Added the definition in cd0b3eb0; `tsc --noEmit` is now clean.

**Comment by ishiland** [comment]
@codex review

File: frontend/src/components/map-plugins/builtin/LegendPlugin.tsx [hunk 4]
```diff
@@ -113,10 +114,17 @@ export function LegendPlugin({ ctx }: { ctx: PluginContext }) {
...
     () => deriveTerrainLegendEntry(ctx.terrainConfig, ctx.layers, { labelKey: 'plugins.legend.terrain3d' }),
     [ctx.terrainConfig, ctx.layers],
   );
+  // Dedup: drop the synthetic entry when the terrain source DEM is ALSO shown as
+  // a per-layer entry (e.g. a visible hillshade of the same dataset), so the
+  // legend doesn't list one DEM twice. Keeps it for the pure-terrain / hidden
+  // case where the synthetic is the only terrain indicator.
+  const terrainEntry = terrainEntryRaw && !terrainSourceIsShownAsLayer(ctx.terrainConfig, legendLayers)
+    ? terrainEntryRaw
+    : null;
```
[the remark this answers is not in the document]
Comment (ishiland) [reply frontend/src/components/map-plugins/builtin/LegendPlugin.tsx]: Fixed — the builder synthetic now gates on the bound DEM being visible (via the shared resolveTerrainSourceLayer BuilderMap uses), so a hidden source = no synthetic, matching BuilderMap clearing terrain. Left the viewer as-is on purpose: useViewerTerrain ignores the toggle, so LayerLegend keeps the synthetic when toggled off and dedups against its own visible set. Added a guard test.

**Comment by ishiland** [comment]
@codex review

**Comment by ishiland** [comment]
Addressed the review-body P2 in 800a3022: dropped the bare `FIX-3-RESOLVER (D-06)` tags from BuilderMap, map-stack, and MapBuilderPage (kept the rationale in plain language), and removed a stale comment that still claimed terrain is active only in Terrain render mode. Comments only. @codex review

**Comment by ishiland** [comment]
@codex review
